### ISSUE TITLE
Overwatch PTR doesn't work

### ISSUE THREAD

**Issue by zekesonxx** [gamax92/wine-overwatch#17]
Ends with the lines:

```

0xE696C954(1544) : Assertion Failed: 0xCA2A37BA  "0x811C9DC5"
fixme:mscoree:CLRCreateInstance not implemented for class {bacc578d-fbdd-48a4-969f-02d932b74634}
```
Log with `WINEDEBUG=fixme-dxgi,fixme-d3d_shader`: [ptr.log.gz](https://github.com/gamax92/wine-overwatch/files/1131808/ptr.log.gz); although it doesn't look very helpful to me.

**Comment by zekesonxx** [comment]
Appears to be class CLRDebugging:

* [clrmd src/Microsoft.Diagnostics.Runtime/ICorDebug/ICorDebugHelpers.cs](https://github.com/Microsoft/clrmd/blob/master/src/Microsoft.Diagnostics.Runtime/ICorDebug/ICorDebugHelpers.cs#L16)
* [coreclr src/inc/metahost.idl](https://github.com/dotnet/coreclr/blob/master/src/inc/metahost.idl#L69)

**Comment by gamax92** [comment]
Personally every PTR has never launched for me, only the live builds, but I'll look into this.

**Comment by zekesonxx** [comment]
This is the first PTR I've ever bothered to download (Doomfist and what have you) so I don't have any previous PTR versions to compare to.

It seems likely that this is because Blizzard probably makes the PTR builds more debugger-friendly (useful crash reports and what have you), so at the very least this hopefully won't affect 1.13 when it goes live.

**Comment by zekesonxx** [comment]
Digging through the Wine code, this doesn't look like it would be too easy to just stub/shim in.

**Comment by zekesonxx** [comment]
PTR update to 1.13.0.0.38044, sort of works? No longer hangs on a black screen, I can login, but it can't successfully connect to a server.

**Comment by zekesonxx** [comment]
So, here's a new one for you.

It works, but only launching *through bnet* (with the autologin). Launching it directly *doesn't* work.

**Comment by zekesonxx** [comment]
Closing since the cause of the crash on 1.13.0.0.37962 is known and the current PTR works.

**Comment by zekesonxx** [comment]
PTR 1.13.0.0.38125, same issue. Blizzard can't seem to decide if they want the class or not..

**Comment by gamax92** [comment]
Considering the mscoree fixme is after OW's Assertion failed message, it very likely has nothing to do with the actual problem.

And yeah, looking back through my logs, this specific `0xE696C954(1544) : Assertion failed: 0xCA2A37BA  "0x811C9DC5"` message is the same one every PTR before has given. Consider 37962 to be a rare case.

**Comment by zekesonxx** [comment]
Yeah that makes sense, it's probably trying to open CLRDebugging to build a crash report or something.

**Comment by gamax92** [comment]
So, wine's winsock2 setsockopt doesn't support `SO_RANDOMIZE_PORT` and infact include/winsock.h is missing a definition for `SO_RANDOMIZE_PORT`, so it ends up setting an error of `WSAENOPROTOOPT` and returning `SOCKET_ERROR`

I threw in a hacky check for `SO_RANDOMIZE_PORT` and made it return 0, and the PTR starts up.

**Comment by gamax92** [comment]
closed by 0f27126df53274197479cb66112067adb7445378

Note that you still have to connect through Battle.net, running the PTR exe directly still doesn't connect.
